### PR TITLE
coverage: Store coverage source regions as `Span` until codegen (take 2)

### PR DESCRIPTION
**Pull request by Zalathar**
This is an attempt to re-land #133418:

> Historically, coverage spans were converted into line/column coordinates during the MIR instrumentation pass.

> This PR moves that conversion step into codegen, so that coverage spans spend most of their time stored as Span instead.

> In addition to being conceptually nicer, this also reduces the size of coverage mappings in MIR, because Span is smaller than 4x u32.

That PR was reverted by #133608, because in some circumstances not covered by our test suite we were emitting coverage metadata that was causing `llvm-cov` to exit with an error (#133606).

---

The implementation here is *mostly* the same, but adapted for subsequent changes in the relevant code (e.g. #134163).

I believe that the changes in #134163 should be sufficient to prevent the problem that required the original PR to be reverted. But I haven't been able to reproduce the original breakage in a regression test, and the `llvm-cov` error message is extremely unhelpful, so I can't completely rule out the possibility of this breaking again.

r? jieyouxu (reviewer of the original PR)

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs
Comment (jieyouxu) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs:144]: Question: with this `continue`, is it possible that we end up with zero available `cov_span`s in total?

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/spans.rs
Comment (jieyouxu) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/spans.rs:83]: :+1:

**Review by jieyouxu — APPROVED** [review]
The implementation looks good to me, and I'm in favoring of making this change. ~~Not having a regression test makes me a bit nervous, but unless we are able to minimize an example that we can use as a regression test (i.e. not a full-blown giant project), I don't think that is super actionable. I think we'll need to find out in nightly if anyone runs into the original problem again, because regardless of making this change or not, I think we do want an MCVE for the original ICE. Furthermore, I think the only way we might be able to minimize one is from more examples.~~ EDIT: we now have a synthetic test example.

Just one question.

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs
Comment (Zalathar) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs:144]: Yes, in rare cases this is possible. I believe that was the cause of the previous bug, because at that time the rest of the code wasn't prepared to deal with that possibility.

Comment (Zalathar) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs:144]: Actually, it occurs to me that even if we don't have an “organic” way to reproduce the bug from source code, I could still rig up a perma-unstable flag that just tells this code to *always* discard every span, and have one test that enables that flag.

Comment (jieyouxu) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs:144]: That seems reasonable to me, yeah. I think it would be beneficial to have such a flag for testing purposes.

Comment (Zalathar) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs:144]: There's already a `-Zcoverage-options` multi-purpose unstable flag, so I'm just going to add it as another value of that flag.

Comment (jieyouxu) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen/covfun.rs:152]: That seems reasonable to me.

**Review by jieyouxu — APPROVED** [review]
The test changes LGTM, you can r=me after PR CI is green.

**Comment by Zalathar** [comment]
I can confirm that if I backport the new test to the point just after #133418 was merged, and do `x test coverage-run`, the llvm-cov tool fails with a “truncated coverage data” error, which is the same error message as in the bug report.

(We don't know for *sure* that it's the same error, because llvm-cov's error reporting for malformed data is really bad, but it's good enough to convince me that the test is doing its job.)

**Comment by jieyouxu** [comment]
It's as good as we're going to get, without nudging llvm-cov to produce better diagnostics.

**Comment by Zalathar** [comment]
🟩

@bors r=jieyouxu

**Pull request merged**
